Update ManualChangeProvider to do a lazy file system scan

jakemac53 opened the pull request:
Should solve https://github.com/flutter/flutter/issues/39696

- Extracted existing file system scanning logic into an `AssetManager` class (in build_runner_core)
- Updated the ManualChangeProvider to use an `AssetManager` to lazily scan the file system instead of using a file watcher and collecting changes.
- Updated the existing test to verify changes were actually picked up by the new build.

natebosch commented on build_runner/lib/src/daemon/daemon_builder.dart:
```diff
@@ -204,17 +206,16 @@ class BuildRunnerDaemonBuilder implements DaemonBuilder {
         .map((data) => WatchEvent(data.type, '${data.id}'))
         .map((change) => [change]);
 
     var changeProvider = daemonOptions.buildMode == BuildMode.Auto
```
Should we pull out all this into a `_createChangeProvider` function? Or maybe move it into the constructor for `AutoChangeProvider`?

jakemac53 replied:
I don't want to move it all into the constructor I don't think (it is a good thing imo that it just accepts a stream of changes right now). I think it is probably fine as is - I could move the inline function to an actual instance method but I don't know that there is really any value there (and I would have more boilerplate for passing in params etc even though its only used locally in this function).

natebosch replied:
Could also consider skipping the local function and moving the whole chunk of code inline where we call the constructor - I'm not sure how ugly that would make things with a large chunk like that in a ternary conditional...

jakemac53 replied:
Ya that gets pretty ugly :/

I also did try extracting it to a separate method but that ends up needing 6 parameters so its pretty boilerplaty compared to what I have now.

natebosch commented on build_runner/lib/src/daemon/daemon_builder.dart:
```diff
@@ -204,17 +206,16 @@ class BuildRunnerDaemonBuilder implements DaemonBuilder {
-    var changes = graphEvents
+            ))
         .map((data) => WatchEvent(data.type, '${data.id}'))
         .map((change) => [change]);
```
If we skip this can we also skip the `.map((changeLists) => changeLists.expand((l) => l).toList()`?

jakemac53 replied:
Oh true it does seem like that, will give it a shot

natebosch commented on build_runner/lib/src/daemon/daemon_builder.dart:
```diff
@@ -204,17 +206,16 @@ class BuildRunnerDaemonBuilder implements DaemonBuilder {
-        ? AutoChangeProvider(changes
+        ? AutoChangeProvider(graphEvents()
             .transform(debounceBuffer(buildOptions.debounceDelay))
             .map((changeLists) => changeLists.expand((l) => l).toList()))
```
Should this get moved to the previous chain of function calls?

jakemac53 replied:
hopefully this can just go away now as you mentioned above, will play with it

natebosch commented on build_runner_core/lib/src/generate/build_definition.dart:
```diff
@@ -64,6 +67,120 @@ class BuildDefinition {
 
+/// Understands how to find all assets relevant to a build as well as compute
+/// updates to those assets.
+class AssetManager {
```
How about `AssetTracker`? Or maybe `AssetDiffer`?

jakemac53 replied:
Renamed AssetTracker

natebosch commented on build_runner_core/lib/src/generate/build_definition.dart:
```diff
@@ -64,6 +67,120 @@ class BuildDefinition {
+
+  AssetManager(this._assetGraph, this._reader, this._targetGraph);
+
+  Future<Map<AssetId, ChangeType>> collectChanges() async {
```
Can we add a doc comment?

grouma commented on build_runner/lib/src/daemon/change_providers.dart:
```diff
@@ -17,21 +22,18 @@ class AutoChangeProvider implements ChangeProvider {
 
-// TODO(grouma) - collect changes through a one time file scan instead of
-// buffering changes from a change stream. This will have better performance
-// on Windows.
```
I think there is a tracking issue that should be linked.

jakemac53 replied:
I can't find one, if you do please close and link it

grouma commented on build_runner/lib/src/daemon/daemon_builder.dart:
```diff
@@ -193,7 +194,8 @@ class BuildRunnerDaemonBuilder implements DaemonBuilder {
 
-    var graphEvents = PackageGraphWatcher(packageGraph,
+    // Only actually used for the AutoChangeProvider.
+    Stream<List<WatchEvent>> graphEvents() => PackageGraphWatcher(packageGraph,
```
Why even create it then? Should we move the conditional above?

jakemac53 replied:
I moved it into a function here so it only gets lazily created if its invoked

grouma commented on build_runner/test/daemon/daemon_test.dart:
```diff
@@ -191,9 +191,9 @@ main() {
-                  print('hello world');
-                }'''),
+main() {
+  print('goodbye world');
```
Dark...

jakemac53 replied:
lol I could go for the exclamation point instead

grouma commented on build_runner_core/lib/src/generate/build_definition.dart:
```diff
@@ -64,6 +67,122 @@ class BuildDefinition {
+
+  /// Checks for and returns any file system changes compared to the current
+  /// state of the asset graph.
+  Future<Map<AssetId, ChangeType>> collectChanges() async {
```
Would be nice to get some unit tests in this package as well.

jakemac53 replied:
done

jakemac53 commented:
Should solve https://github.com/flutter/flutter/issues/39696

- Extracted existing file system scanning logic into an `AssetManager` class (in build_runner_core)
- Updated the ManualChangeProvider to use an `AssetManager` to lazily scan the file system instead of using a file watcher and collecting changes.
- Updated the existing test to verify changes were actually picked up by the new build.
